atuin-client/src/record/sqlite_store.rs

```diff
@@ -122,29 +123,75 @@ impl Store for SqliteStore {
     }
 
     async fn get(&self, id: RecordId) -> Result<Record<EncryptedData>> {
-        let res = sqlx::query("select * from records where id = ?1")
-            .bind(id.0.as_simple().to_string())
+        let res = sqlx::query("select * from store where store.id = ?1")
+            .bind(id.0.as_hyphenated().to_string())
             .map(Self::query_row)
             .fetch_one(&self.pool)
             .await?;
 
         Ok(res)
     }
 
-    async fn len(&self, host: HostId, tag: &str) -> Result<u64> {
-        let res: (i64,) =
-            sqlx::query_as("select count(1) from records where host = ?1 and tag = ?2")
-                .bind(host.0.as_simple().to_string())
+    async fn last(&self, host: HostId, tag: &str) -> Result<Option<Record<EncryptedData>>> {
+        let res =
+            sqlx::query("select * from store where host=?1 and tag=?2 order by idx desc limit 1")
+                .bind(host.0.as_hyphenated().to_string())
                 .bind(tag)
+                .map(Self::query_row)
                 .fetch_one(&self.pool)
+                .await;
+
+        match res {
+            Err(sqlx::Error::RowNotFound) => Ok(None),
+            Err(e) => Err(eyre!("an error occured: {}", e)),
+            Ok(record) => Ok(Some(record)),
+        }
+    }
+
+    async fn first(&self, host: HostId, tag: &str) -> Result<Option<Record<EncryptedData>>> {
+        self.idx(host, tag, 0).await
+    }
+
+    async fn len(&self, host: HostId, tag: &str) -> Result<Option<u64>> {
+        let last = self.last(host, tag).await?;
+
+        if let Some(last) = last {
+            return Ok(Some(last.idx + 1));
+        }
+
+        return Ok(None);
+    }
+
+    async fn next(
+        &self,
+        host: HostId,
+        tag: &str,
+        idx: RecordIdx,
+        limit: u64,
+    ) -> Result<Vec<Record<EncryptedData>>> {
+        let res =
+            sqlx::query("select * from store where idx >= ?1 and host = ?2 and tag = ?3 limit ?4")
```

Comment:
Seems strange to me that `next` would include the current idx field. I assume you add 1 elsewhere

Reply:
ok, I see how this is used during sync_upload. I think I might suggest a different name, or at least a different doc comment because `/// Get the record that follows this record` is wrong.

Maybe `get_page` and `/// Get a record page starting at the current record index`